src/main/java/io/temporal/internal/sync/POJOWorkflowInterfaceMetadata.java

```diff
@@ -0,0 +1,247 @@
+    private final WorkflowMethodType type;
+
+    EqualsByNameType(String name, WorkflowMethodType type) {
+      this.name = name;
+      this.type = type;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      EqualsByNameType that = (EqualsByNameType) o;
+      return com.google.common.base.Objects.equal(name, that.name) && type == that.type;
+    }
+
+    @Override
+    public int hashCode() {
+      return com.google.common.base.Objects.hashCode(name, type);
+    }
+  }
+
+  private POJOWorkflowMethodMetadata workflowMethod;
+  private final Map<Method, POJOWorkflowMethodMetadata> methods = new HashMap<>();
+
+  public static POJOWorkflowInterfaceMetadata newInstance(Class<?> anInterface) {
+    if (!anInterface.isInterface()) {
+      throw new IllegalArgumentException("Not an interface: " + anInterface);
+    }
+    WorkflowInterface annotation = anInterface.getAnnotation(WorkflowInterface.class);
+    if (annotation == null) {
+      throw new IllegalArgumentException(
+          "Missing requied @WorkflowInterface annotation: " + anInterface);
+    }
+    POJOWorkflowInterfaceMetadata result = new POJOWorkflowInterfaceMetadata(anInterface);
+    if (result.methods.isEmpty()) {
+      throw new IllegalArgumentException(
+          "Interface doesn't contain any methods: " + anInterface.getName());
+    }
+    return result;
+  }
+
+  static POJOWorkflowInterfaceMetadata newImplementationInterface(Class<?> anInterface) {
+    return new POJOWorkflowInterfaceMetadata(anInterface);
+  }
+
+  private POJOWorkflowInterfaceMetadata(Class<?> anInterface) {
+    Map<EqualsByMethodName, Method> dedupeMap = new HashMap<>();
+    getWorkflowInterfaceMethods(anInterface, dedupeMap);
+  }
+
+  public Optional<POJOWorkflowMethodMetadata> getWorkflowMethod() {
+    return Optional.ofNullable(workflowMethod);
+  }
+
+  public Optional<String> getWorkflowType() {
+    if (workflowMethod == null) {
+      return Optional.empty();
+    }
+    return Optional.of(workflowMethod.getName());
+  }
+
+  public POJOWorkflowMethodMetadata getMethodMetadata(Method method) {
+    POJOWorkflowMethodMetadata result = methods.get(method);
+    if (result == null) {
+      throw new IllegalArgumentException("Unknown method: " + method);
+    }
+    return result;
+  }
+
+  public List<POJOWorkflowMethodMetadata> getMethodsMetadata(WorkflowMethodType type) {
+    List<POJOWorkflowMethodMetadata> result = new ArrayList<>();
+    for (POJOWorkflowMethodMetadata methodMetadata : this.methods.values()) {
+      if (methodMetadata.getType() == type) {
+        result.add(methodMetadata);
+      }
+    }
+    return result;
+  }
+
+  public List<POJOWorkflowMethodMetadata> getMethodsMetadata() {
+    return new ArrayList<>(this.methods.values());
+  }
+
+  private void initWorkflowInterface(Class<?> anInterface) {
+    if (!anInterface.isInterface()) {
```

Comment:
nit: init and constructor have some similar code that might be able to be turned into util

Reply:
Good catch. It is actually leftover of me moving this code to constructor :). Removing this method completely.